pkg/await/core_pod.go

```diff
@@ -0,0 +1,293 @@
+		if !isMap || containerStatus["ready"] == true {
+			continue
+		}
+
+		// Best effort attempt to get name of container. (This should always succeed and if it
+		// doesn't, it's not worth crashing the provider over).
+		rawName := containerStatus["name"]
+		var name string
+		name, _ = rawName.(string)
+
+		rawWaiting, isWaiting := openapi.Pluck(containerStatus, "state", "waiting")
+		waiting, isMap := rawWaiting.(map[string]interface{})
+		if isWaiting && rawWaiting != nil && isMap {
+			pc.checkWaitingContainer(name, waiting)
+		}
+
+		//
+		// TODO: terminated
+		//
+	}
+
+	// Exhausted our knowledge of possible error states for Pods. Return.
+}
+
+func (pc *podChecker) checkWaitingContainer(name string, waiting map[string]interface{}) {
+	rawReason, hasReason := waiting["reason"]
+	reason, isString := rawReason.(string)
+	if !hasReason || !isString || reason == "" || reason == "ContainerCreating" {
+		return
+	}
+
+	rawMessage, hasMessage := waiting["message"]
+	message, isString := rawMessage.(string)
+	if !hasMessage || !isString {
+		return
+	}
+
+	// Image pull error has a bunch of useless junk at the beginning of the error message. Try to
+	// remove it.
+	imagePullJunk := "rpc error: code = Unknown desc = Error response from daemon: "
+	message = strings.TrimPrefix(message, imagePullJunk)
+
+	pc.containerErrors[reason] = append(pc.containerErrors[reason], message)
+}
+
+func (pc *podChecker) clearErrors() {
+	pc.podScheduledErrors = map[string]string{}
+	pc.containerErrors = map[string][]string{}
+}
+
+func (pc *podChecker) errorMessages() []string {
+	messages := []string{}
+	for reason, message := range pc.podScheduledErrors {
+		messages = append(messages, fmt.Sprintf("Pod unscheduled: [%s] %s", reason, message))
+	}
+
+	for reason, errors := range pc.containerErrors {
+		for _, message := range errors {
+			messages = append(messages, fmt.Sprintf("[%s] %s", reason, message))
+		}
+	}
+	return messages
+}
+
+// --------------------------------------------------------------------------
+
+// POD AWAITING. Routines for waiting until a Pod has been initialized correctly.
+
+// --------------------------------------------------------------------------
+
+type podInitAwaiter struct {
+	podChecker
+	config createAwaitConfig
+}
+
+func makePodInitAwaiter(c createAwaitConfig) *podInitAwaiter {
+	return &podInitAwaiter{
+		config:     c,
+		podChecker: *makePodChecker(),
+	}
+}
+
+func (pia *podInitAwaiter) Await() error {
+	//
+	// We succeed when `.status.phase` is set to "Running".
+	//
+
+	podWatcher, err := pia.config.clientForResource.Watch(metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "Could set up watch for Pod object '%s'",
+			pia.config.currentInputs.GetName())
+	}
+	defer podWatcher.Stop()
+
+	// Create timeout channel.
+	timeoutCh := make(chan struct{})
+	go func() {
+		time.Sleep(5 * time.Minute)
+		timeoutCh <- struct{}{}
+	}()
+
+	return pia.await(podWatcher, timeoutCh)
```

Comment:
`pia.await(podWatcher, time.Tick(5 * time.Minute))` should work too

Reply:
I see you use this pattern a bunch of places in here, would it be obtrusive to make timeoutCh a `chan Time` instead of a `chan struct{}`?

Reply:
Let's do that, but the linter informs me that we want to use `time.NewTicker` instead, or else we leak the the ticker.